### PR TITLE
feat: linking and project listing commands

### PR DESCRIPTION
**Pull request by aorumbayev**
# Proposed Changes

## Command Line Interface Enhancements
- **List Command Support**: Introduces a simple `ls` command for browsing currently available projects in the workspace, enhancing navigation and project visibility.
- **Link Command Support**: Simplifies the use of the `algokit generate` command by allowing interactive invocation or exporting typed clients from sibling contract projects, streamlining the development workflow.

## Deployment and Bootstrap Enhancements
- **Deployment Flexibility**: Enables deployment from the workspace root by specifying a project name with `algokit deploy`, allowing for more centralized management of projects.
- **Bootstrap Enhancements**: Adds the capability to use `algokit bootstrap` with filters by project type or name, facilitating targeted initialization of projects.

## Run Command Improvements
- **Environment File Specification**: Enhances the `run` command to support an `env_file` field, prioritizing the loading of custom environment files. This feature defaults to `.env` files if present and ultimately loads the global context, ensuring environmental configurations are accurately applied.
- **Filtering Capabilities**: Introduces filtering by project-type and name in `run` commands when orchestrated from the root, allowing for more refined execution control.
- **Multiple Commands Support**: Adds the ability to specify multiple `commands` under `run`, offering a cross-platform alternative to Unix-like command chaining. This ensures commands are valid and resolved before execution, enhancing command flexibility and reliability.

## GitHub Actions and Workspace Management
- **GitHub Files Resolution**: Addresses the handling of `.github` files for standalone projects. With the workspace flag enabled, GitHub actions within standalone projects are not discoverable by default -> hence the fix. Implements a move operation for any `.github` folder content at the project's root to the workspace's root `.github` folder post-initialization. This change resolves a bug from the 1.12.x release that required manual movement of standalone production preset templates to the workspace structure.

## TODO

- [x] - snap shot tests for link and list commands, edge cases for deploy and boostrap 
- [x]  - documentation 

Related template PRs:

https://github.com/algorandfoundation/algokit-react-frontend-template/pull/27
https://github.com/algorandfoundation/algokit-puya-template/pull/7
https://github.com/algorandfoundation/algokit-base-template/pull/3
https://github.com/algorandfoundation/algokit-fullstack-template/pull/22

### CODE REVIEW

File: src/algokit/cli/project/deploy.py [hunk 1]
```diff
@@ -147,54 +220,75 @@ def deploy_command(  # noqa: PLR0913
...
+        projects = get_project_configs(project_type=ProjectType.CONTRACT)
+
+        for project in projects:
+            project_name = project.get("project", {}).get("name", None)
+
+            if not project_name:
+                click.secho("WARNING: Skipping an unnamed project...", fg="yellow")
+                continue
+            if project_names and project_name not in project_names:
+                click.secho(
```
[the remark this answers is not in the document]
Comment (aorumbayev) [reply src/algokit/cli/project/deploy.py]: @neilcampbell refactored the method loading the configs to include filtering by name, so this conditional is removed completely

File: src/algokit/cli/project/deploy.py [hunk 2]
```diff
@@ -78,6 +80,58 @@ def _ensure_environment_secrets(
...
+            msg = f"No generic deploy command specified in '{ALGOKIT_CONFIG}' file."
+        else:
+            msg = (
+                f"Deploy command for '{environment_name}' is not specified in '{ALGOKIT_CONFIG}' file, "
+                "and no generic command."
+            )
+        raise click.ClickException(msg)
+    resolved_command = resolve_command_path(config.command)
+    logger.info(f"Using deploy command: {' '.join(resolved_command)}")
+    # TODO: [future-note] do we want to walk up for env/config?
```
[the remark this answers is not in the document]
Comment (aorumbayev) [reply src/algokit/cli/project/deploy.py:108]: this is an old future note, don't think we want to do that at this point given the capabilities around orchestration, will remove note

File: .python-version [hunk 3]
```diff
@@ -0,0 +1 @@
+3.12
```
[the remark this answers is not in the document]
Comment (aorumbayev) [reply .python-version]: typo, removed file, poetry creates it automatically when you run poetry env use commands

File: src/algokit/cli/init.py [hunk 4]
```diff
@@ -429,6 +439,60 @@ def _maybe_git_init(project_path: Path, *, use_git: bool | None, commit_message:
...
+            except shutil.Error as e:
+                logger.debug(f"Skipping move of {source_file} to {target_file}: {e}")
+
+    # if source dir does not contain any files or only contains empty folders
+    # then remove it
+
+    if all(not p.is_file() for p in source_dir.rglob("*")):
+        shutil.rmtree(source_dir)
+        click.echo(
+            "The contents of your `.github` folder have been moved to the workspace's root `.github` folder. "
```
[the remark this answers is not in the document]
Comment (aorumbayev) [reply src/algokit/cli/init.py]: refined by logging this only in debug and default echo being simplified to your one-sentence suggestion

**Comment by aorumbayev** [comment]
@PhearZero @joe-p @Loedn @iskysun96 - available from branch for any further feedback. Try test this out use:

```
pipx install git+https://github.com/algorandfoundation/algokit-cli@feat/orchestration-linking
```

Please note only react/puya and their fullstack combo are currently updated to latest, beaker is to be updated after puya is reviewed, for tealscript ill add a simple build/test run declarations into toml @joe-p but for the rest refer to official templates to get tealscript template codebase up to date with latest conventions (notion of presets, tests, ci/cd, code tours, generators etc).

File: src/algokit/cli/init.py [hunk 5]
```diff
@@ -429,6 +439,52 @@ def _maybe_git_init(project_path: Path, *, use_git: bool | None, commit_message:
...
+                logger.debug(f"Skipping move of {source_file} to {target_file}: {e}")
+
+    # if source dir does not contain any files or only contains empty folders
+    # then remove it
+    if all(not p.is_file() for p in source_dir.rglob("*")):
+        shutil.rmtree(source_dir)
+        logger.debug(f"No files found in .github folder after merge. Removing `.github` directory at {source_dir}...")
+        return
+
+    click.echo(
```
Comment (neilcampbell) [inline src/algokit/cli/init.py:483]: It appears that this will only render if there are files found in the project .github folder.
To me this should be printed always at the start and we should replace this message with something like below:

`Failed to move all files within your project's `.github` folder to the workspace root. Please review any files that remain in your project's `.github` folder and manually include in the root `.github` directory as required.`

**Review by neilcampbell — APPROVED** [review]

**Review by robdmoore — APPROVED** [review]
